### PR TITLE
Improve http client error handling

### PR DESCRIPTION
**Pull request by Patrik-Stas**
- Improve error messages
- Propagate error details when mapping agency error into VcxError

Signed-off-by: Patrik Stas <patrik.stas@absa.africa>

### CODE REVIEW

File: agency_client/src/httpclient.rs [hunk 1]
```diff
@@ -53,21 +53,28 @@ pub async fn post_message(body_content: &Vec<u8>, url: &str) -> AgencyClientResu
             .send()
             .await
             .map_err(|err| {
-                AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, format!("Could not connect {:?}", err))
+                let err_msg = format!("HTTP Client could not connect with ${}, err: {}", url, err.to_string());
```
Comment (mirgee) [inline agency_client/src/httpclient.rs:56]: Not nodejs, the `$` won't help :) (`${}`)

File: agency_client/src/httpclient.rs [hunk 2]
```diff
@@ -53,21 +53,28 @@ pub async fn post_message(body_content: &Vec<u8>, url: &str) -> AgencyClientResu
...
-                Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, format!("Agency response could not be read.")))
+    let content_length = response.content_length();
+    let response_status = response.status();
+    match response.text().await {
+        Ok(payload) => {
+            if response_status.is_success() {
+                Ok(payload.into_bytes())
+            } else {
+                let err_msg = format!("POST {} failed due non-success HTTP status: {}, response body: {}", url, response_status.to_string(), payload);
+                error!("{}", err_msg);
```
Comment (mirgee) [inline agency_client/src/httpclient.rs:69]: dtto (I really dislike this pattern)

File: agency_client/src/httpclient.rs [hunk 3]
```diff
@@ -53,21 +53,28 @@ pub async fn post_message(body_content: &Vec<u8>, url: &str) -> AgencyClientResu
...
+                Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, err_msg))
             }
         }
-    } else {
-        Ok(response.text().await
-            .or(Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, "could not read response")))?.into())
+        Err(error) => {
+            let err_msg = format!("POST {} failed because response can not be decoded as utf-8 text, HTTP status: {}, content-length header: {:?}, error: {:?}", url, response_status.to_string(), content_length, error);
+            error!("{}", err_msg);
```
Comment (mirgee) [inline agency_client/src/httpclient.rs:75]: dtto

File: agency_client/src/httpclient.rs [hunk 4]
```diff
@@ -53,21 +53,28 @@ pub async fn post_message(body_content: &Vec<u8>, url: &str) -> AgencyClientResu
...
-            Err(_) => {
-                Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, format!("Agency response could not be read.")))
+    let content_length = response.content_length();
+    let response_status = response.status();
+    match response.text().await {
+        Ok(payload) => {
+            if response_status.is_success() {
+                Ok(payload.into_bytes())
+            } else {
+                let err_msg = format!("POST {} failed due non-success HTTP status: {}, response body: {}", url, response_status.to_string(), payload);
```
Comment (mirgee) [inline agency_client/src/httpclient.rs:68]: "due to"

File: agency_client/src/httpclient.rs [hunk 5]
```diff
@@ -53,21 +53,28 @@ pub async fn post_message(body_content: &Vec<u8>, url: &str) -> AgencyClientResu
...
+                error!("{}", err_msg);
+                Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, err_msg))
             }
         }
-    } else {
-        Ok(response.text().await
-            .or(Err(AgencyClientError::from_msg(AgencyClientErrorKind::PostMessageFailed, "could not read response")))?.into())
+        Err(error) => {
+            let err_msg = format!("POST {} failed because response can not be decoded as utf-8 text, HTTP status: {}, content-length header: {:?}, error: {:?}", url, response_status.to_string(), content_length, error);
```
Comment (mirgee) [inline agency_client/src/httpclient.rs:74]: "...response could not be decoded into utf-8, HTTP status ..."

**Review by mirgee — CHANGES_REQUESTED** [review]

File: agency_client/src/httpclient.rs (hunk 5 above)
Comment (Patrik-Stas) [reply agency_client/src/httpclient.rs:74]: Modified but kept "as utf-8", we are not decoding "into", rather decoding "<something>"

**Review by mirgee — APPROVED** [review]

**Pull request merged**